### ISSUE TITLE
[Discuss] Operationalize container exit events to avoid container engine memory leak

### ISSUE THREAD

**Issue by incertum** [falcosecurity/libs#1227]
**Describe the bug**

Memory leak in containers cache in container engine when trying to operationalize container exit events in this PR https://github.com/falcosecurity/libs/pull/1228.

Initially wrongly attributed to introducing mutex lock, because of reading this https://devblogs.microsoft.com/oldnewthing/20201112-00/?p=104444

Need to make additional adjustments to the PR to ensure we keep the correct behavior.

**Comment by gnosek** [comment]
Before I take a deep look at it, this line looks wrong:

https://github.com/falcosecurity/libs/commit/33b47e92cf97b1df162c282a48906c7bbb0ab4ef#diff-0b9f04fd797523bf7465503b515f794972719a6c83c696f1b749b7ba1b9169c0R87

@incertum, can you try the following patch to see if it changes anything?

```
- containers->erase(it++);
+ it = containers->erase(it);
```

**Comment by incertum** [comment]
No that is not fixing it I actually tried it at first. This is a great post that explains why you need delayed destruction because of the lock https://devblogs.microsoft.com/oldnewthing/20201112-00/?p=104444.

Btw yes that line definitely needs changing to the secure and best coding practice anyways, but as said it won't fix the leak.

**Comment by gnosek** [comment]
I read that post and while I agree with the premise, I don't see how it would cause a memory leak. A deadlock, sure, but I just can't see how we'd end up with leaking memory because of the lock. Can you describe such a scenario to me?

(if you look at the rationale https://devblogs.microsoft.com/oldnewthing/20201111-00/?p=104439 linked from your post, you'll also see it boils down to introducing a deadlock).

Does reverting https://github.com/falcosecurity/libs/commit/33b47e92cf97b1df162c282a48906c7bbb0ab4ef fix the leak?

> Btw yes that line definitely needs changing to the secure and best coding practice anyways, but as said it won't fix the leak.

I'm pretty sure that line is UB as it is and I can totally imagine it causing a memory leak or a crash (it was buggy before too, but the UB may have manifested in a different way).

**Comment by incertum** [comment]
Maybe the easiest is could you try on master yourself? Containers are just not being removed even when changing to `it = containers->erase(it);`?

I did not yet revert https://github.com/falcosecurity/libs/commit/33b47e92cf97b1df162c282a48906c7bbb0ab4ef, but I suppose it would then work again as it's then a normal unordered map again. I figured I rather ask you here first before investing too much time.

**Comment by gnosek** [comment]
So what's the debug output you added and what are you getting?

For me it looks fine even without the iterator fix (I know I'm lucky here):

The container appears (not yet present in the container table but present in a threadinfo):
```
07-26 13:08:27.964286 Finding containers still in use
07-26 13:08:27.964346 Container 7548c946a23c in use
07-26 13:08:27.964455 Scanning 0 containers
07-26 13:08:27.964457 Done scanning containers, 0 active
```

The container is alive:
```
07-26 13:08:28.964571 Finding containers still in use
07-26 13:08:28.964622 Container 7548c946a23c in use
07-26 13:08:28.964724 Scanning 1 containers
07-26 13:08:28.964726 Container 7548c946a23c still in use
07-26 13:08:28.964727 Done scanning containers, 1 active
```

etc. etc.

The container exits:
```
07-26 13:08:37.965799 Finding containers still in use
07-26 13:08:37.965967 Scanning 1 containers
07-26 13:08:37.965969 Container 7548c946a23c no longer in use
07-26 13:08:37.965994 Done scanning containers, 0 active
```

The `%d active` comes from `containers->size()` so the map really is empty then.



> Maybe the easiest is could you try on master yourself? Containers are just not being removed even when changing to `it = containers->erase(it);`?

It sure was easiest, it involved only a several hour debugging session involving kernel mode null derefs from eBPF :P 

> I did not yet revert [33b47e9](https://github.com/falcosecurity/libs/commit/33b47e92cf97b1df162c282a48906c7bbb0ab4ef), but I suppose it would work again as it's then a normal unordered map.

It's *still* a normal unordered map behind the mutex :man_shrugging: 

> I figured I rather ask you here first before investing too much time.

Same.

> "Memory leak" here means containers are not removed from the map, we keep growing memory as new containers come up on the node.

Yes, that would make perfect sense when we're mishandling the iterator.

This is what I'm running:

```
diff --git a/userspace/libsinsp/container.cpp b/userspace/libsinsp/container.cpp
index 71ae7e06c..fe8e38772 100644
--- a/userspace/libsinsp/container.cpp
+++ b/userspace/libsinsp/container.cpp
@@ -78,19 +78,23 @@ bool sinsp_container_manager::remove_inactive_containers()
 
 		threadinfo_map_t* threadtable = m_inspector->m_thread_manager->get_threads();
 
+		g_logger.format(sinsp_logger::SEV_CRITICAL, "Finding containers still in use");
 		threadtable->loop([&] (const sinsp_threadinfo& tinfo) {
 			if(!tinfo.m_container_id.empty())
 			{
 				containers_in_use.insert(tinfo.m_container_id);
+				g_logger.format(sinsp_logger::SEV_CRITICAL, "Container %s in use", tinfo.m_container_id.c_str());
 			}
 			return true;
 		});
 
 		auto containers = m_containers.lock();
+		g_logger.format(sinsp_logger::SEV_CRITICAL, "Scanning %d containers", containers->size());
 		for(auto it = containers->begin(); it != containers->end();)
 		{
 			if(containers_in_use.find(it->first) == containers_in_use.end())
 			{
+				g_logger.format(sinsp_logger::SEV_CRITICAL, "Container %s no longer in use", it->first.c_str());
 				sinsp_container_info::ptr_t container = it->second;
 				for(const auto &remove_cb : m_remove_callbacks)
 				{
@@ -100,9 +104,11 @@ bool sinsp_container_manager::remove_inactive_containers()
 			}
 			else
 			{
+				g_logger.format(sinsp_logger::SEV_CRITICAL, "Container %s still in use", it->first.c_str());
 				++it;
 			}
 		}
+		g_logger.format(sinsp_logger::SEV_CRITICAL, "Done scanning containers, %d active", containers->size());
 	}
 
 	return res;
diff --git a/userspace/libsinsp/examples/test.cpp b/userspace/libsinsp/examples/test.cpp
index 69775ab37..ccc4d8933 100644
--- a/userspace/libsinsp/examples/test.cpp
+++ b/userspace/libsinsp/examples/test.cpp
@@ -337,6 +337,9 @@ int main(int argc, char** argv)
 	sinsp inspector;
 	dump = plaintext_dump;
 
+	g_logger.set_severity(sinsp_logger::SEV_NOTICE);
+	g_logger.add_stderr_log();
+
 #ifndef _WIN32
 	parse_CLI_options(inspector, argc, argv);
 
diff --git a/userspace/libsinsp/settings.h b/userspace/libsinsp/settings.h
index c79618a57..6167e03e2 100644
--- a/userspace/libsinsp/settings.h
+++ b/userspace/libsinsp/settings.h
@@ -47,7 +47,7 @@ limitations under the License.
 //
 // How often the container table is scanned for inactive containers
 //
-#define DEFAULT_INACTIVE_CONTAINER_SCAN_TIME_S 30
+#define DEFAULT_INACTIVE_CONTAINER_SCAN_TIME_S 1
 
 //
 // How often the users/groups tables are scanned for deleted users/groups
```

Can you give it  a try and let me know what it logs?

**Comment by incertum** [comment]
Thanks @gnosek :pray: now this is starting to get really strange ... let me explain.

First I used my branch from this PR https://github.com/falcosecurity/libs/pull/1228.
Added you changes on top with exception of changing g_logger init to `g_logger.add_stdout_log();`.
Left the change `it = containers->erase(it);`. Simulated `docker run -it ubuntu:latest bash` and then exited.


When running a condition that does not produce any outputs, I get what you get.

```
[x]$ sudo ./libsinsp/examples/sinsp-example -b driver/bpf/probe.o -f "(evt.type=container and proc.name=bogus)" -o "*%evt.dir %proc.name %proc.exepath %container.id %container.start_ts %container.duration %container.image.repository"
-- Try to open: 'bpf' engine.
driver/bpf/probe.o
-- Engine 'bpf' correctly opened.
-- Start capture
07-26 18:18:25.982359 Finding containers still in use
07-26 18:18:25.982401 Scanning 1 containers
07-26 18:18:25.982404 Container 94d01972a073 no longer in use
07-26 18:18:25.982521 Done scanning containers, 0 active
07-26 18:18:26.990748 Finding containers still in use
07-26 18:18:26.991435 Scanning 0 containers
07-26 18:18:27.965773 Finding containers still in use
07-26 18:18:27.966407 Scanning 0 containers
07-26 18:18:28.972020 Finding containers still in use
07-26 18:18:28.972676 Scanning 0 containers
07-26 18:18:29.977574 Finding containers still in use
07-26 18:18:29.978279 Scanning 0 containers
07-26 18:18:30.983600 Finding containers still in use
07-26 18:18:30.984298 Scanning 0 containers
07-26 18:18:31.988900 Finding containers still in use
07-26 18:18:31.989576 Scanning 0 containers
07-26 18:18:32.966386 Finding containers still in use
07-26 18:18:32.966986 Scanning 0 containers
```

Now when I produce some new output logs I get these symptoms that suggest we are not removing the containers:

```
[x]$ sudo ./libsinsp/examples/sinsp-example -b driver/bpf/probe.o -f "(evt.type=container)" -o "*%evt.dir %proc.name %proc.exepath %container.id %container.start_ts %container.duration %container.image.repository"
-- Try to open: 'bpf' engine.
driver/bpf/probe.o
-- Engine 'bpf' correctly opened.
-- Start capture
{"container.duration":442307454,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":">","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:33.366055 Finding containers still in use
07-26 18:19:33.366097 Scanning 1 containers
07-26 18:19:33.366101 Container df88a49732b5 no longer in use
07-26 18:19:33.366226 Done scanning containers, 0 active
{"container.duration":25347366701,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:34.374737 Finding containers still in use
07-26 18:19:34.375470 Scanning 1 containers
07-26 18:19:34.375493 Container df88a49732b5 no longer in use
07-26 18:19:34.375875 Done scanning containers, 0 active
{"container.duration":26347777099,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:35.350305 Finding containers still in use
07-26 18:19:35.350944 Scanning 1 containers
07-26 18:19:35.350967 Container df88a49732b5 no longer in use
07-26 18:19:35.351311 Done scanning containers, 0 active
{"container.duration":27348238519,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:36.357573 Finding containers still in use
07-26 18:19:36.358256 Scanning 1 containers
07-26 18:19:36.358280 Container df88a49732b5 no longer in use
07-26 18:19:36.358607 Done scanning containers, 0 active
{"container.duration":28348653233,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
```

Hmm what should we look into next? Maybe now I am thinking the issue is elsewhere? Glad we are having the discussion here first! WDYT?

**Comment by incertum** [comment]
Thanks @gnosek :pray: now this is starting to get really strange ... let me explain.

First I used my branch from this PR https://github.com/falcosecurity/libs/pull/1228.
Added you changes on top with exception of changing g_logger init to `g_logger.add_stdout_log();`.
Left the change `it = containers->erase(it);`. Simulated `docker run -it ubuntu:latest bash` and then exited.


When running a condition that does not produce any outputs, I get what you get.

```
[x]$ sudo ./libsinsp/examples/sinsp-example -b driver/bpf/probe.o -f "(evt.type=container and proc.name=bogus)" -o "*%evt.dir %proc.name %proc.exepath %container.id %container.start_ts %container.duration %container.image.repository"
-- Try to open: 'bpf' engine.
driver/bpf/probe.o
-- Engine 'bpf' correctly opened.
-- Start capture
07-26 18:18:25.982359 Finding containers still in use
07-26 18:18:25.982401 Scanning 1 containers
07-26 18:18:25.982404 Container 94d01972a073 no longer in use
07-26 18:18:25.982521 Done scanning containers, 0 active
07-26 18:18:26.990748 Finding containers still in use
07-26 18:18:26.991435 Scanning 0 containers
07-26 18:18:27.965773 Finding containers still in use
07-26 18:18:27.966407 Scanning 0 containers
07-26 18:18:28.972020 Finding containers still in use
07-26 18:18:28.972676 Scanning 0 containers
07-26 18:18:29.977574 Finding containers still in use
07-26 18:18:29.978279 Scanning 0 containers
07-26 18:18:30.983600 Finding containers still in use
07-26 18:18:30.984298 Scanning 0 containers
07-26 18:18:31.988900 Finding containers still in use
07-26 18:18:31.989576 Scanning 0 containers
07-26 18:18:32.966386 Finding containers still in use
07-26 18:18:32.966986 Scanning 0 containers
```

Now when I produce some new output logs I get these symptoms that suggest we are not removing the containers:

```
[x]$ sudo ./libsinsp/examples/sinsp-example -b driver/bpf/probe.o -f "(evt.type=container)" -o "*%evt.dir %proc.name %proc.exepath %container.id %container.start_ts %container.duration %container.image.repository"
-- Try to open: 'bpf' engine.
driver/bpf/probe.o
-- Engine 'bpf' correctly opened.
-- Start capture
{"container.duration":442307454,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":">","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:33.366055 Finding containers still in use
07-26 18:19:33.366097 Scanning 1 containers
07-26 18:19:33.366101 Container df88a49732b5 no longer in use
07-26 18:19:33.366226 Done scanning containers, 0 active
{"container.duration":25347366701,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:34.374737 Finding containers still in use
07-26 18:19:34.375470 Scanning 1 containers
07-26 18:19:34.375493 Container df88a49732b5 no longer in use
07-26 18:19:34.375875 Done scanning containers, 0 active
{"container.duration":26347777099,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:35.350305 Finding containers still in use
07-26 18:19:35.350944 Scanning 1 containers
07-26 18:19:35.350967 Container df88a49732b5 no longer in use
07-26 18:19:35.351311 Done scanning containers, 0 active
{"container.duration":27348238519,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
07-26 18:19:36.357573 Finding containers still in use
07-26 18:19:36.358256 Scanning 1 containers
07-26 18:19:36.358280 Container df88a49732b5 no longer in use
07-26 18:19:36.358607 Done scanning containers, 0 active
{"container.duration":28348653233,"container.id":"df88a49732b5","container.image.repository":"ubuntu","container.start_ts":1690395548000000000,"evt.dir":"<","proc.exepath":"container:df88a49732b5","proc.name":"container:df88a49732b5"}
```

In summary, there are much deeper intricacies in the libsinsp state that re-add the containers here just because of trying to operationalize the exit events ... as on master when running it it's also ok.

Hence need to check on the parser better ... :/ will update the ticket title accordingly.

**Comment by incertum** [comment]
@gnosek pushed to https://github.com/falcosecurity/libs/pull/1228 (also adjusting the better practice `it` handling to at least get a bit more value out of this discussion).

Probably should also add better logging while here as well ... basically your suggestions.

Could we turn this into a design discussion?

What I don't understand is why we add a container to the cache on `notify_new_container` and then again buried in `parse_container_json_evt` which probably is not wise to overload a function responsible for parsing with this functionality. Likely it fixes something, but it makes me wonder if a cleanup/redesign wouldn't be more sustainable?

**Comment by incertum** [comment]
@gnosek pushed to https://github.com/falcosecurity/libs/pull/1228 (also adjusting the better practice `it` handling to at least get a bit more value out of this discussion).

Probably should also add better logging while here as well ... basically your suggestions.

Could we turn this into a design discussion?

So, we add a container to the cache on `notify_new_container` and then again buried in `parse_container_json_evt` which probably is not wise to overload a function responsible for parsing with this functionality. 
Because I suppose now it has the updated / parsed info?
It makes me wonder if a cleanup/redesign wouldn't be more sustainable?
